### DISCUSSION TITLE
[Chapter4.6-4.10] 트랜지스터에서 CPU로, 이보다 더 중요한 것은 없다.

### DISCUSSION THREAD

**Discussion by ella-yschoi** [elegant-functional-2023/secrets-of-computer#13] [General]
### 감상평을 남겨주세요! 📝

- 범위: Chapter 4.6-4.10
- 기한: 7/27(토) 오전 11시까지

**Comment by yunseorim1116** [comment]
다시 한 번 혼공컴 읽었을때의 기록이 새록새록 떠올랐다. RISC, CISC를 읽으면서 컴퓨터 공학과 친구에게 “이걸 꼭 알아야할까?” 하고 물어봤는데 “당연한거 아니니..?” 하고 답이 왔었던 기억이 났다. 여전히 내 머리는 기억속에서 리셋을 시켜버렸다. 하지만 이 책의 설명이 좀 더 자세해서 다음엔 꼭꼭 기억할수 있을 것 같다. 이번 파트.. 왜 이렇게 생소하고 길고 어려웠을까..? … 개인적으로 제일 어려웠던것같다. 특히 4.9부터는 우다다다 쏟아지는 느낌(TT) 컴퓨터랑 친해지는건 어렵다................복습을 많이 해야겠다고 느꼈다.

[정리링크](https://nifty-period-b9a.notion.site/7-abef5b82b6cb431bb51b86252c59af5f?pvs=25)

**Reply by sunhpark42** [reply]
"당연한거 아니니..?"

**Reply by ddongule** [reply]
당연하긴 머가당연해!!!!!!!!!!!!!!!!

**Comment by ella-yschoi** [comment]
## 감상평

- 역사적 관점에서 CPU를 이해하니 호흡이 빠르고 재밌었습니다.
- 하지만 4.9장은 넘나 자세한 내용을 다루고 있어 저도 조금 어려웠읍니다,,
- 이번 챕터를 읽고나니 이런 기사가 다르게 보이네요! [애플도 MS도 'x86' 떠난다…인텔의 해법은?](https://ddaily.co.kr/page/view/2021082015150829781)

<br/>

## 4.6장: CPU 진화론(상): 복잡 명령어 집합의 탄생

- 책에 나온 CISC는 비유하자면 암산을 하기 때문에 머리에서 열이 나지만 (발열), 대신 종이를 아낄 수 있고 (메모리 차지 X)
- 반면 RISC(Reduced Insctuction Set Computer)는 종이에 써서 발열은 없지만, 종이를 낭비(메모리 차지) 한다고 하네요.
- 어쨌든 이번 장에서 말하는 역사는 '의미상 간격(semantic gap)'을 좁히려는 과정들 같다는 생각을 했습니다.
- 즉, 사용자가 원하는 작업과 문제해결방식, 그리고 컴퓨터가 실제로 행하는 저수준 연산 간의 차이를 좁히는 것.

<br/>

## 4.7장: CPU 진화론 (하): 축소 명령어 집합의 탄생

- 예전에 설탕 값이 비쌌을 때 비해 요즘은 설탕을 사용한 식품이 흔해지고 식품 산업의 발전이 빨라졌듯이,
- 컴퓨터 과학 쪽에서도 메모리 용량 대비 가격이 급격히 떨어지며 컴파일 기술 등 프로그래밍 기술 또한 빨라진 것 같습니다.
- 파레토의 법칙이 여기서도 나올 줄이야..!
- '복잡 명령어 집합에 대한 반성 → 축소 명령어 집합의 등장' 이 스토리에서 '반성' 이라는 단어가 흥미로웠어요.
- 역시 기술의 발전은 이전 기술의 한계로부터 비롯되는 것을 또 한번 느꼈습니다.

<br/>

## 4.8장: CPU 진화론 (하): 절체절명의 위기에서의 반격!

- 엎치락 뒤치락 복잡 명령어 집합 vs. 축소 명령어 집합 재격돌!
- 어쨌든 마이크로 명령어와 하이퍼스레딩 덕분에 둘은 원만한 합의를 이룰 수 있었군요,,
- CISC vs. RISC 간 상업적 전쟁 내용에서는 상대적으로 최근인 애플의 M1칩 개발 이야기가 나와서 반가웠습니다.

<br/>

## 4.9장: CPU, 스택과 함수 호출, 시스템 호출, 스레드 전환, (헥헥..) 인터럽트 처리 통달하기

- 4.9장은 앞에서 배운 내용을 한번 더 짚고가는 느낌이네요. 뭔가 같은 이야기를 계속 반복해서 설명하는 것 같은데...
- 특히 앞 장에서는 레지스터에 대해 간단하게 짚고 넘어간 느낌이었는데, 이번에 자세하게 설명해주어 좋았습니다.
- 하지만 어려웡.........

<br/>

**Reply by hanana1253** [reply]
헥헥..ㅋㅋ

**Reply by sunhpark42** [reply]
"재밌었습니다."
"하지만 어려웡........."

**Comment by iborymagic** [comment]
- 4.9를 제외하고는 CPU의 역사에 관한 내용이 대부분이라 술술 읽을 수 있었던 것 같습니다..
- 근데 4.9는 갑자기 내용이 복잡해져서 사실 아직 내용을 완전히 이해하지 못했읍니다.. 특히 마지막에 스레드 전환 부분은 뭐라는지 도통 알아먹지를 못하겠네요.
- 책에서는 RISC와 CISC의 차이가 거의 없어져간다고 했지만, 요즘 인텔과 애플칩의 격차가 점점 벌어지는 걸 생각해보면 결국 근본적인 성능 차이는 좁힐 수 없는건가 싶기도 합니다.. 허허
- CPU의 역사를 읽으면서.. 인텔, AMD, 애플, 테슬라 등 각 회사들의 전성기에는 짐 켈러 아저씨가 있었다는 얘기가 떠올랐읍니다.. 자세히 알지는 못하지만 멋있는 아저씨..

**Reply by sunhpark42** [reply]
"짐 켈러" 인텔, AMD, 애플, 테슬라 등 각 회사들의 전성기에는 그가 있었다.

**Comment by ddongule** [comment]
- CPU에 대해 알아가는 시간..이었..습니다.. 
- 아직 알아가고.. 있습니다...

**Reply by hanana1253** [reply]
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 알아가는 단계군요..

**Reply by ella-yschoi** [reply]
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ천천히... 알아가보아요...

**Comment by hanana1253** [comment]
지난 주에 좀 쉽다고 생각했는데 사망플래그였구나... 이번에는 이해가 안됨ㅎ.. 
루 샤오펑(저자) 아조시가 스토리텔링 맛깔나게 잘 하시네.. 중국사람이라 그런가 삼국지 읽는 것 같은 기분이 들었다.
오늘은 왠지 호스트가 안 걸릴 것 같은 느낌적인 느낌... 4.9만 아니면 돼!

맥북에 뭐 깔 때마다 x86이나 ARM 같은거 이해 못해서 검색해서 깔곤 했는데 이젠 그래도 아는 척할 수 있겠다

## 4.6 
CPU & 명령어 집합의 발전도 일련의 정반합 과정을 거친다는 생각이 듭니다.

- 프로그램을 직접 어셈블리어로 작성하기 때문에 풍부한 명령어 집합 필요 , 컴파일러의 미성숙, 저장장치 용량제한 -> 복잡 명령어 집합 등장
- CPU 명령어 집합은 모두 직접 hardwired -> 새로운 명령어 추가 시 CPU 설게 및 디버깅 복잡도 증가 -> 명령어에 포함된 연산을 더 간단한 명령어로 구성된 작은 프로그램으로 정의하여 CPU에 저장 -> 마이크로코드 
- 마이크로코드 버그 수정은 일반적인 버그보다 훨씬 어렵고 마이크로코드의 설계는 트랜지스터 소모량 많음 -> 축소 명령어 집합 탄생


## 4.7

> p.329 복잡 명령어 집합에서 성능을 향상시키는 것으로 여겨지는 명령어는 실제로 CPU 내부에서 마이크로코드에 의해 가로막힌다는 것입니다.
> p.330 마이크로코드가 없으면 컴파일러에서 생성된 기계 명령어의 CPU 제어 능력이 크게 향상됩니다.

뭔말이지... T_T CPU내부에 마이크로코드가 자리를 차지하고있어서 성능향상에 한계가 있다는 얘긴가? 자리 뿐 아니라 코드 자체로도 성능이슈가 있단 얘긴가?

전에 나왔던 파이프라인 공정이 뜬금없이 튀어나온게 아니었구나... 짧고 균등한 시간을 사용하는 명령어들이 잔뜩 있을 때 어떻게 효과적으로 처리량을 늘리는지 이해가 되었다.

> 컴퓨터의 CPU 또는 디지털회로가 일정한 속도로 작동하기 위해서는 일정한 간격으로 전기적 진동(pulse)을 공급 받아야 한다. 즉 CPU를 비롯한 컴퓨터의 모든 부품들은 특정한 신호에 맞추어 동작을 하는데, 이 특정한 신호를 가리키는 말이 바로 '클럭'이라는 단어이다. (네이버 블로그)

클럭이라는 단위가 있구나. 세상엔 참 내가 모르는 다양한 것들이 있다는 것을 깨달았다.

## 4.8 

x86이 쉽게 죽지 않는 걸 보면서 선빵 필승과 생태계(보급률 및 대중성)의 확보라는 중요성이 기술 시장이라 해서 피해가지 않는 느낌. 아니 오히려 기술 시장에서 더 중요한 것 같기도.. 그 와중에 또 스마트폰 신시장은 몸집 작고 재빠른 ARM이 낼름 먹었다는 것도 재밌다.

**Reply by hanana1253** [reply]
맥북 배터리가 없어서 일단 4.8까지 기록했음. 진짜임. 4.9는 지금 하고있어요

**Reply by hanana1253** [reply]
## 4.9

분명 이전 챕터들에서 봤던 내용들인데 왜이렇게 생소하지.. 

> p.346 산술 연산이 포함된 명령어는 수행 중에 자리 올림수가 발생하거나 넘침이 발생할 수 있습니다. 이런 정보는 상태 레지스터에 저장됩니다.

저장해서 어따 쓰는겨...? 에러 내는 건가..?

**Reply by sunhpark42** [reply]
> 오늘은 왠지 호스트가 안 걸릴 것 같은 느낌적인 느낌... 4.9만 아니면 돼!

예측 성공!

**Reply by iborymagic** [reply]
동희) 클럭을 몰라??

**Reply by ddongule** [reply]
숨쉬듯이 무례하네요

**Comment by thgr8ganzi** [comment]
* 서림닝의 CISC, RISC 의 중요성에 대한 감상평을 읽고 다시 읽어봤습니다.
* 이번장은 개인적으로 읽기가 조금 힘들었습니다. 조금 흥미가 떨어지는 부분이 있었습니다.
* 4.9 부분 같은 경우는 지금까지의 복습과 얼추 비슷하면서 좀 더 설명해주는 게 좋았습니다. 레지스터의 종류가 많네요

**Reply by bigsaigon333** [reply]
- CISC, RISC 이름만 어렴풋이 기억하고 있었는데, 자세히 알게되어서 좋았습니다.
- x86, ARM ... M1 으로 이어지는
흐름이 흥미롭네요 ㅎㅎ

**Reply by ella-yschoi** [reply]
동희님!!ㅋㅋㅋㅋㅋ 댓글로 다셨어요 크크

**Comment by sa02045** [comment]
1. CPU의 발전과정을 알 수 있었습니다. x86,  CISC, RISC 생소한 단어를 습득했습니다
2. 기술의 발전에는 상업적, 비용적 이유가 필연적이라는 것을 느꼈습니다. 비싼 메모리 -> 복합 명령어 집합 등장. 저렴해진 메모리 -> 축수 명령어 집합 등장
3. 하이퍼스레딩처럼 처음듣는 단어가 있어서 흥미로웠습니다."커널 상태 스택"도 있었군요
4.  전반적으로 밑바닥에서 CPU가 어떻게 동작하고 발전해왔는지, 레지스터와 사용자 스택 + 커널 상태 스택을 구분해서 생각할 수 있게되어서 새로운 개념을 배웠고 좋았습니다

- CPU 입장에서 "hello world" 프로그램이든 "포토샵"같은 대규모 프로그램이든 차이가 없다!
   - 단지 처리해야할 명령어가 적거나 많거나
- 명령어 집합: CPU가 할 수 있는 일
  - 명령어 집합의 명령어를 CPU로 보내면 CPU는 해당 명령어를 수행한다.
  - 명령어는 프로그래머가 프로그래밍에 사용한다.

- 명령어집합의 종류: CISC(복합 명령어 집합)
- 폰 노이만 구조: "프로그램이나 데이터나 동일하다. 모두 메모리 안에 저장된다."

- 왜 복합 명령어 집합이 먼저 발전했는가?
   - 메모리 공간 때문!
   - 옛날에는 사용할 수 있는 메모리가 무척이나 적었다.
   - 코드 공간을 절약해야한다!   
   - 모두 직접 연결(hardwired) 방식으로 명령어 인출, 해독, 실행 단계를 직접 제어한다.
 
- 복합 명령어 집합의 문제
   - 하드웨어는 변경이 어렵다! 소프트웨어는 변경이 쉽다!
   - 마이크로코드 의 등장

마이크로코드 설계의 문제점
   - 마이크로 코드 설계가 트랜지스터를 많이 소모한다.
   - 마이크로코드의 버그를 수정하는 것이 매우 어렵다.

- 축소 명령어 집합의 탄생
   - 컴파일러의 발전 + 메모리 용양의 발전 => 프로그래머가 더이상 어셈블리어로 프로그램을 작성하지 않아도 된다. 고오급 언어로 작성한다!
   - 복잡성을 단순하게 + 컴파일러가 더 많은 CPU 제어권을 가지도록 
   - 마치 함수형 프로그래밍처럼 함수를 분리하고 각자의 역할을 할 수 있도록
- 복잡 명령어 집합 진영도 `인터페이스`를 도입하여 복잡 명령어 집합을 사용해도 축소 명령어 집합을 유사하게 동작하도록 만들었다

- `하이퍼 스레딩`
   - CPU가 한 번에 두개 스레드의 명령어 흐름을 처리하는 것
   - 파이프라인 기술으로

**Comment by chloe-codes1** [comment]
## 감상평

- CISC 와 RISC 방식의 차이점과 단점을 보완하기 위해 등장한 하이퍼 스레딩, 그리고 apple silicon 의 등장 등 CPU 가 발전한 흐름을 알 수 있어 좋았다
- 전에 하이퍼 스레딩을 공부했었는데, CISC 와 RISC 개념을 정립하고 다시 공부한 내용을 읽어보니 이해가 더 잘되는 것 같다!

하이퍼 스레딩 정리: https://chloe-codes1.gitbook.io/til/os/os-basics/02_what_is_hyper-threading

**Reply by ddongule** [reply]
![image](https://github.com/user-attachments/assets/8a2b014e-440c-4462-a57a-a9fbbab9345e)
다앙~

**Comment by sunhpark42** [comment]
- 이번 4장 자체는 확실히 역사책 같은 느낌이 있다.
- 교수님이 이분이었다면 난 A+ 을 받을 수 있었을까..? 느껴지는 덕후의 향기로 보았을때 이분 수업이면 B0 였을 듯.
- 확실히 컴퓨터 구조, 스레딩 이런거 보다보면 운영체제 다시 공부하고 싶어지는 것 같다.

**Reply by ddongule** [reply]
상대평가이기때문에 A+가능할거같아요

**Reply by sunhpark42** [reply]
저희 절평이라 ㅠ

**Comment by bigsaigon333** [comment]
- CISC, RISC 이름만 어렴풋이 기억하고 있었는데, 자세히 알게되어서 좋았습니다.
- x86, ARM ... M1 으로 이어지는
흐름이 흥미롭네요 ㅎㅎ
